server/src/main/java/org/opensearch/plugins/IdentityPlugin.java

```diff
@@ -38,8 +38,8 @@ public interface IdentityPlugin {
      * Gets a subject corresponding to the passed plugin that can be utilized to perform transport actions
      * in the plugin system context
      *
-     * @param plugin The corresponding plugin
+     * @param pluginInfo The corresponding pluginInfo
      * @return Subject corresponding to the plugin
      */
-    PluginSubject getPluginSubject(Plugin plugin);
+    PluginSubject getPluginSubject(PluginInfo pluginInfo);
```

Comment:
Why do we need to use `PluginInfo`? Plugin does not need to be told what action he is allowed to use: it comes from the its permissions policy.

Reply:
Its lighter-weight than Plugin and its the vehicle that carries the actions that a plugin may request to perform with its PluginSubject.

The security plugin would need these 2 pieces of information to form the PluginSubject:

1. Plugin Identifier (by convention it would be the canonical class name of the plugin)
2. Request Actions <- This is what this PR is introducing. i.e. The security plugin is requesting permissions to write to the security auditlog index with its pluginSubject.

Reply:
Got it, this is indeed the conceptual problem (at least I overlooked). I think this pull request is a bit ahead of its time, here is my reasoning:
 - the core has no notion of permissions whatsoever
 - as of today, this is purely and only `security` plugin feature, and
 - any alternative (identity?) plugin implementation would have known nothing about permissions

I still think that the idea to expose the plugin permissions is sound but we need a way to formalize the permissions first and (may be?) make core / other plugins aware of them (basically, do what we have done with `Subject` etc). 

Why I think that is important, when we install any plugin right now, the security policy will be enforced no matter what. With permissions, as it stand now, this is purely hint, nothing will happen if `security` plugin is not installed / disabled (or any other its replacement is present).

@peternied sorry to dragging you here, but I think the subject of permissions what discussed before, right?

Reply:
FWIW I think presenting this information to an administrator installing a plugin empowers them with more information about what actions a plugin will perform in the cluster regardless if the security plugin is installed or not.

I left a comment [here](https://github.com/opensearch-project/security/issues/4439#issuecomment-2174505670) about complications of putting system index protection directly in the core. The biggest challenge with that is that operators may need to be able to perform invasive actions on system indices and do so by presenting the admin certificate which is another feature of the security plugin.

Reply:
JSM Permissions actually work the same way. A cluster operator is prompted with the permissions on install even if they run opensearch with JSM disabled.

Reply:
> JSM Permissions actually work the same way. A cluster operator is prompted with the permissions on install even if they run opensearch with JSM disabled.

I believe it is not possible to disable ISM (except tests), it is installed unconditional (please correct me if I am wrong here).

Reply:
> FWIW I think presenting this information to an administrator installing a plugin empowers them with more information about what actions a plugin will perform in the cluster regardless if the security plugin is installed or not.

This is right, and I agree with that. The question is how to do that in a way that if administrator says "Looks good" but nothing is enforced?

Reply:
You're right its not possible to disable outside of the [tests](https://github.com/opensearch-project/OpenSearch/blob/main/test/framework/src/main/java/org/opensearch/bootstrap/BootstrapForTesting.java#L128). JSM being removed in JDK 24 is a disruptive change. lmk if I can help with review on any items.

Reply:
Please join the discussion https://github.com/opensearch-project/OpenSearch/issues/1687

Reply:
Hey @reta, I opened a draft PR in core to add a new sandbox module for system index protection: https://github.com/opensearch-project/OpenSearch/pull/16695

In its current state it is a crude implementation and would need to be expanded upon to match the functionality that the security plugin provides.

I wanted to open a draft to demonstrate some of the challenges in moving system index protection to the core. Ultimately, I think it is a good idea so that they are protected regardless if the security plugin is installed and enabled. 

One of the biggest challenges is generic ActionRequest -> indices() resolution in an action filter. The way this works in the security plugin is the [IndexResolverReplacer.getOrReplaceAllIndices()](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java#L640-L832).

Reply:
> Hey @reta, I opened a draft PR in core to add a new sandbox module for system index protection: #16695

Apologies for the delay @cwperks , I will try to find the time next week to look at it, thanks !